my_enums.rb

```diff
@@ -1,60 +1,138 @@
 # frozen_string_literal: true
 
-# rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
-
 module Enumerable #:nodoc: all
   def my_each
+    return to_enum unless block_given?
+
     length.times do |n|
       current = self[n]
       yield(current)
     end
+    self
   end
 
   def my_each_with_index
-    length.times do |n|
-      index = n
-      current = self[n]
-      yield(index, current)
+    return to_enum unless block_given?
+
+    my_each do |n|
+      yield(n, index(n))
     end
+    self
   end
 
   def my_select
     emp_arr = []
+    return to_enum unless block_given?
+
     my_each do |n|
       emp_arr << n if yield(n)
     end
     emp_arr
   end
 
   def my_all?
-    emp_arr = []
-    my_each do |n|
-      emp_arr << n if yield(n)
+    if block_given?
+      begin
+        if yield.is_a? Class
+          my_each do |n|
+            return false unless n.instance_of? yield
+          end
+          true
+        elsif yield.instance_of? Regexp
+          my_each do |n|
+            next if n =~ yield
+
+            return false
+          end
+          true
+        end
+      rescue NoMethodError
+        my_each do |n|
+          next if yield(n)
+
+          return false
+        end
+        true
+      end
+    else
+      return false if empty?
+
+      my_each do |n|
+        return false if n.nil? || n == false
+      end
+      true
     end
-    emp_arr == self
   end
 
   def my_any?
-    tester = false
-    my_each do |n|
-      tester = true if yield(n)
-      break if tester
+    if block_given?
+      begin
+        if yield.is_a? Class
+          my_each do |n|
+            return true if n.instance_of? yield
+          end
+          false
+        elsif yield.instance_of? Regexp
+          my_each do |n|
+            return true if n =~ yield
+          end
+          false
+        end
+      rescue NoMethodError
+        my_each do |n|
+          return true if yield(n)
+        end
+        false
+      end
+    else
+      return false if empty?
+
+      my_each do |n|
+        return true unless n.nil? || n == false
+      end
+      false
     end
-    tester
   end
 
   def my_none?
-    tester = true
-    my_each do |n|
-      tester = false if yield(n)
-      break if tester == false
+    if block_given?
+      begin
+        if yield.is_a? Class
+          my_each do |n|
+            return false if n.instance_of? yield
+          end
+          true
+        elsif yield.instance_of? Regexp
+          my_each do |n|
+            return false if n =~ yield
+          end
+          true
+        end
+      rescue NoMethodError
+        tester = true
+        my_each do |n|
+          tester = false if yield(n)
+          break if tester == false
+        end
+        tester
+      end
+    else
+      return true if empty?
+
+      my_each do |n|
+        return false if n == true
+      end
+      true
     end
```

Comment:
* [ ] The two statements below ought to return ```true``` and ```false``` as it does in ruby's ```none?``` method  
```
 %w{ant bear cat}.my_none?(/d/)                        #=> true
 [1, 3.14, 42].my_none?(Float)                         #=> false 
```